mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs

```diff
@@ -110,92 +110,174 @@ public static void ChangeWrapperHandleStrength (object otp, bool fIsWeak)
 			throw new NotImplementedException ();
 		}
 
-		[MethodImplAttribute(MethodImplOptions.InternalCall)]
-		internal extern static void copy_to_unmanaged (Array source, int startIndex,
-							       IntPtr destination, int length);
+		unsafe internal static void copy_to_unmanaged (Array source, int startIndex,
+		       IntPtr destination, int length)
+		{
+			copy_to_unmanaged_fixed (source, startIndex, destination, length, null);
+		}
 
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
-		internal extern static void copy_from_unmanaged (IntPtr source, int startIndex,
-								 Array destination, int length);
+		unsafe private extern static void copy_to_unmanaged_fixed (Array source, int startIndex,
+		       IntPtr destination, int length, void* fixed_source_element);
+
+		static private bool skip_fixed (System.Array array, int startIndex)
+		{
+			// In particular, we see length == 0 && startIndex == array.Length, and fixed fails.
+			return startIndex < 0 || startIndex >= array.Length;
+		}
 
-		public static void Copy (byte[] source, int startIndex, IntPtr destination, int length)
+		unsafe internal static void copy_to_unmanaged (byte[] source, int startIndex, IntPtr destination, int length)
 		{
-			copy_to_unmanaged (source, startIndex, destination, length);
+			if (skip_fixed (source, startIndex))
+				copy_to_unmanaged (source, startIndex, destination, length);
```

Comment:
@jaykrell  This doesn't call the `System.Array` override. it just loops.

Reply:
I see, will fix this morning.